blog/2022-11-21-golang-memory-dump.md

```diff
@@ -0,0 +1,187 @@
+
+This is extremely useful when you want to debug your application since they can contain debugging tools that you don't to have in your application container.
+
+In this post I thought we could go trough how to profile a running container in Kubernetes.
+
+<!-- truncate -->
+
+## Application
+
+To get started lets use a very simple test [application](https://github.com/polarsignals/pprof-example-app-go) written by the team over at [Polar Signals](https://www.polarsignals.com/).
+
+They have been kind enough to publish a container image and a Kubernetes deployment that we will use.
+The application is built in such a way that it will continue to use memory and the Kubernetes yaml don't contain any request nor limit so don't run this for a long time or your system will probably OOM.
+
+Run the application on Kubernetes 1.23 or higher.
+
+```shell
+kubectl apply -f https://raw.githubusercontent.com/polarsignals/pprof-example-app-go/main/manifests/deployment.yaml
+```
+
+For starters lets have a look at the pprof http endpoint using curl.
+A simple way of doing so is to create a Kubernetes service and reach the pod from another container.
+
+```shell
+# Expose the pprof deployment
+kubectl expose deployment pprof-example-app-go --port=8080
+```
+
+Create a curl pod and look at the data inside our app.
+
+```shell
+kubectl run -i -t curl --image=curlimages/curl:latest /bin/sh
+```
+
+You will be sent directly in to the container and you can run something like:
+
+```shell
+curl http://pprof-example-app-go:8080/debug/pprof/allocs?debug=1
+```
+
+This will show you an output that look something like this:
+
+```shell
+heap profile: 4: 47357952 [5473: 554303632] @ heap/1048576
+1: 46882816 [1: 46882816] @ 0x697585 0x470ce1
+#	0x697584	main.allocMem+0xa4	/home/brancz/src/github.com/polarsignals/pprof-example-app-go/main.go:65
+
+1: 212992 [1: 212992] @ 0x4e3b6e 0x4e3e6c 0x6974c5 0x470ce1
+#	0x4e3b6d	log.(*Logger).Output+0x38d	/usr/local/go/src/log/log.go:180
+#	0x4e3e6b	log.Println+0x6b		/usr/local/go/src/log/log.go:329
+#	0x6974c4	main.calculateFib+0xc4		/home/brancz/src/github.com/polarsignals/pprof-example-app-go/main.go:55
+
+1: 204800 [1: 204800] @ 0x4fb68f 0x4fb256 0x503d3d 0x502c97 0x4f6d5e 0x4f6c92 0x4d22a5 0x4d2625 0x4d70b1 0x4cf3d2 0x4e3e3f 0x6974c5 0x470ce1
+#	0x4fb68e	math/big.nat.make+0x5ee		/usr/local/go/src/math/big/nat.go:69
+#	0x4fb255	math/big.nat.sqr+0x1b5		/usr/local/go/src/math/big/nat.go:595
+```
+
+Sadly this output isn't the easiest to read so why not use pprof and while we are at it why not use ephemeral containers.
+
+## Debug container/ephemeral container
+
+Before doing my research my plan was to write about how we attached an volume so we could
+save data locally inside the ephemeral container and then copy it out of the container on to our client and
+do show some nice flame graphs. But apparently it's not supported to attach volumes, it's not even supported to reach the existing volumes on the pod.
```

Comment:
Instead of many small comments, here is a suggested alternative. Pick what parts you agree with.

"My original plan for this blog post was to attach a new volume to the ephemeral container so we could save data locally  and then copy it out to our client and do show some nice flame graphs. But apparently it's not supported to attach volumes to the ephemeral container. The ephemeral container cannot even reach the existing volumes on the pod you attach to."